src/storage-client/src/controller.rs

```diff
@@ -2405,6 +2439,110 @@ where
+
+        // Add the frontiers known to the storage controller.
+        for (collection_id, collection) in &self.state.collections {
+            // Some of the storage frontiers are not advanced by us but by our users, which might
+            // provide them to us in `frontiers` too. Ensure we don't end up duplicates because of
+            // this.
+            frontiers.remove(collection_id);
+            if let Some(time) = collection.write_frontier.clone().into_option() {
+                push_frontier(*collection_id, time);
+            }
+        }
+        for (export_id, export) in &self.state.exports {
+            // Some of the storage frontiers are not advanced by us but by our users, which might
+            // provide them to us in `frontiers` too. Ensure we don't end up duplicates because of
+            // this.
+            frontiers.remove(export_id);
+            if let Some(time) = export.write_frontier.clone().into_option() {
+                push_frontier(*export_id, time);
+            }
+        }
+
+        // Add the provided frontiers.
+        for (object_id, time) in frontiers {
+            push_frontier(object_id, time);
+        }
+
+        // Cancel out superfluous retractions.
+        consolidate(&mut updates);
+
+        self.append_to_managed_collection(
+            self.state.introspection_ids[&IntrospectionType::Frontiers],
+            updates,
+        )
+        .await;
+    }
+
+    async fn record_replica_frontiers(
+        &mut self,
+        frontiers: BTreeMap<ReplicaId, Vec<(GlobalId, Self::Timestamp)>>,
+    ) {
+        let old_rows = &mut self.replica_frontiers_rows;
+        let mut updates = Vec::with_capacity(old_rows.len());
+
+        // Retract existing rows.
+        updates.extend(old_rows.drain(..).map(|row| (row, -1)));
+
+        let mut push_frontier = |replica_id, object_id: GlobalId, time: T| {
+            // TODO(#18377): Make replica IDs `NewReplicaId`s throughout the code.
+            let replica_id = NewReplicaId::User(replica_id);
+            let row = Row::pack_slice(&[
+                Datum::String(&replica_id.to_string()),
+                Datum::String(&object_id.to_string()),
+                Datum::MzTimestamp(mz_repr::Timestamp::from_bytes(time.encode())),
+            ]);
+            updates.push((row.clone(), 1));
+            old_rows.push(row);
+        };
+
+        // Add the frontiers known to the storage controller.
+        for (collection_id, collection) in &self.state.collections {
+            let Some(cluster_id) = collection.cluster_id() else { continue };
+            let Some(replica_id) = self.state.replicas.get(&cluster_id) else { continue };
+            let Some(time) = collection.write_frontier.clone().into_option() else { continue };
+            push_frontier(*replica_id, *collection_id, time);
+        }
+        for (export_id, export) in &self.state.exports {
+            let Some(cluster_id) = export.cluster_id() else { continue };
```

Comment:
I'm not sure why this method (which is inherent to `ExportState` and not a trait impl returns an `Option<_>`. If you look at its implementation it unconditionally returns a cluster id. Could you change the method and remove the `Option<_>` wrapper since you're here?

Reply:
Will do! My guess is that this was done for consistency with `CollectionState::cluster_id`.